geowebcache/sqlite/src/main/java/org/geowebcache/sqlite/SqliteConnectionManager.java

```diff
@@ -0,0 +1,425 @@
+/**
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * <p/>
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * <p/>
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author Nuno Oliveira, GeoSolutions S.A.S., Copyright 2016
+ */
+package org.geowebcache.sqlite;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.File;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Manages the connections to sqlite databases files taking care of the concurrent access.
+ * The concurrent access are managed by JVM if two JVMs access the same database file the
+ * result is unpredictable.
+ */
+public final class SqliteConnectionManager {
+
+    private static Log LOGGER = LogFactory.getLog(SqliteConnectionManager.class);
+
+    private final ConcurrentHashMap<File, PooledConnection> pool = new ConcurrentHashMap<>();
+
+    private volatile boolean stopPoolReaper = false;
+
+    public SqliteConnectionManager(SqliteConfiguration configuration) {
+        this(configuration.getPoolSize(), configuration.getPoolReaperIntervalMs());
+    }
+
+    SqliteConnectionManager(long poolSize, long poolReaperIntervalMs) {
+        Utils.info(LOGGER, "Initiating connection poll: [poolSize='%d', poolReaperIntervalMs='%d'].",
+                poolSize, poolReaperIntervalMs);
+        // let's load the sqlite driver
+        try {
+            Class.forName("org.sqlite.JDBC");
+        } catch (Exception exception) {
+            throw Utils.exception(exception, "Error initiating sqlite driver.");
+        }
+        // computing some values used by the pool reaper
+        double poolSizeThreshold = poolSize * 0.9;
+        double connectionsToRemove = poolSize * 0.1;
+        // starting the connection pool reaper
+        new Thread(() -> {
```

Comment:
Also, since the eviction happens at a certain interval, it means the number of files open concurrently go above threshold for a certain amount of time. This should be documented.
I am left wondering why a guava cache was not used instead, it would have done the job of keeping the size at max all times without needing a separate thread: https://github.com/google/guava/wiki/CachesExplained
Not saying you have to rewrite it, if it works it works...